### DISCUSSION TITLE
Breaking change - Adding the Async suffix

### DISCUSSION THREAD

**Discussion by julienasp** [vkhorikov/CSharpFunctionalExtensions#466] [General]
I wanted to add an Execute extension method for func ( practically a sugar syntax for Map ) but in doing so, i notice that the signature was clashing with the Execute ValueTask since it is not suffixed with Async.

So i was wondering why all theses Task/ValueTask async method are not suffixed with Async.

I can do the work to fix all of those everywhere by adding the Async suffix, but it would be a huge breaking changes. But considering that those changes were added less than 4 months ago. It shouldn't be that bad for those who are using those new extension methods.

so what do you think? can i do it?

**Comment by hankovich** [comment]
Hi, I'm also a big fan of `Async` suffix and I use it everywhere except for controllers and tests (because action methods and test methods are called by external framework).

But I think that in this library there's no need for that suffix since `CSharpFunctionalExtensions` provides a dsl for `Result`s and `Maybe`s.
When you are familiar with the provided ubiquitous language than it's extremely easy to read the method chains like

```csharp 
return GetAsync(id)
   .Map(...)
   .Bind(...)
   .TapError(...)
   .Compensate(...);
```

And it's useful that you can more or less understand what is going on just by reading this chain, without digging in to `...` methods. You can focus on business logic and not on the c# concepts. And later you can change an implementation of any `...` method (make is return `Task<T>`, not just `T`) without modification of the chain.

I think the vast majority of used extensions in my projects are called on tasks, so I'll have to write a lot of `Async` suffixes even in places where it feels weird. For example

```csharp
return GetVersionAsync(id)
   .MapAsync(version => version++); // map function is sync, but since it's called on a task (so it is so called `async left` overload), we have to await this task, so we have to add this suffix to `Map`s name
```

So I think this change will break a nice domain language and just add suffuxes without any practical needs.

**Comment by vkhorikov** [comment]
`Async` suffix is essentially an implementation detail. The only reason to put it to a method name is to avoid signature collision in overloaded methods. If there's no collision overload (like in our case), then there's no need for this suffix.
